src/combinatorics/partition.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+
+import combinatorics.composition
+import data.nat.parity
+import tactic.apply_fun
+
+/-!
+# Partitions
+
+A partition of a natural number `n` is a way of writing `n` as a sum of positive integers, where the order
+does not matter: two sums that differ only in the order of their summands are considered the same
+partition. This notion is closely related to that of a composition of `n`, but in a composition of
+`n` the order does matter.
+
+## Main functions
+
+* `p : partition n` is a structure, made of a multiset of integers which are all positive and
+  add up to `n`.
+
+## Implementation details
+
+The main motivation for this structure and its API is to show Euler's partition theorem, and
+related results.
+
+The representation of a partition as a multiset is very handy as multisets are very flexible and
+already have a well-developed API.
+
+## Tags
+
+Partition
+
+## References
+
+<https://en.wikipedia.org/wiki/Partition_(number_theory)>
+-/
+
+
+variables {α : Type*}
+
+open multiset nat
+open_locale big_operators
+
+/-- A partition of `n` is a multiset of positive integers summing to `n`. -/
+@[ext, derive decidable_eq] structure partition (n : ℕ) :=
+(parts : multiset ℕ)
+(parts_pos : ∀ {i}, i ∈ parts → 0 < i)
+(parts_sum : parts.sum = n)
+
+namespace partition
+
+/-- A composition induces a partition (just convert the list to a multiset). -/
+def of_composition (n : ℕ) (c : composition n) : partition n :=
+{ parts := c.blocks,
+  parts_pos := λ i hi, c.blocks_pos hi,
+  parts_sum := by rw [multiset.coe_sum, c.blocks_sum] }
+
+lemma of_composition_surj {n : ℕ} : function.surjective (of_composition n) :=
+begin
+  rintro ⟨b, hb₁, hb₂⟩,
+  rcases quotient.exists_rep b with ⟨b, rfl⟩,
+  refine ⟨⟨b, λ i hi, hb₁ hi, _⟩, partition.ext _ _ rfl⟩,
+  simpa using hb₂
+end
+
+/--
+Given a multiset which sums to `n`, construct a partition of `n` with the same multiset, but
+without the zeros.
+-/
+def of_sums (n : ℕ) (l : multiset ℕ) (hl : l.sum = n) : partition n :=
```

Comment:
```suggestion
def of_sums {n : ℕ} (l : multiset ℕ) (hl : l.sum = n) : partition n :=
```
? Or is there a reason it's convenient to keep this explicit?

Reply:
Yeah it's useful to keep this explicit because when I used this in a proof, I usually wanted to generate the new goal of giving the value of the sum. I don't have a strong objection to making it implicit, but I think it's easier to use as explicit (and I'm happy to add a comment saying such if you think that's a good idea)